### PR TITLE
Allow reloading of search time analyzers

### PR DESCRIPTION
**Pull request by cbuescher**
Currently changing resources (like dictionaries, synonym files etc...) of search
time analyzers is only possible by closing an index, changing the underlying
resource (e.g. synonym files) and then re-opening the index for the change to
take effect.

This PR adds a new API endpoint that allows triggering reloading of certain
analysis resources (currently token filters) that will then pick up changes in
underlying file resources. To achieve this we introduce a new type of custom
analyzer (ReloadableCustomAnalyzer) that uses a ReuseStrategy that allows
swapping out analysis components. Custom analyzers that contain filters that are
markes as "updateable" will automatically choose this implementation. This PR
also adds this capability to synonym token filters for use in search time
analyzers.

Relates to #29051

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java]: Summary of comments from #42669:

Alan expressed the ides to have ReloadableCustomAnalyzer extend CustomAnalyzer, which is something I originally did in the first WIP PR (#40782) which Jim was somewhat against. He suggested pulling out a common base class. 

My take on this, that I implemented in a follow up commit here, is to make both Analyzers implement a common interface that allows uniform access to the AnalyzerComponents without having to change the CustomAnalyzer API much.

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/SynonymAnalyzerTests.java
Comment (cbuescher) [inline modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/SynonymAnalyzerTests.java]: Summary of comments from #42669:

* Alan expressed the wish for thread-safety test here (e.g. create TokenStreams in multiple threads, trigger a reload in a separate thread, and check that no tokenstream ends up with 'mixed' synonyms)
* I asked for some clarification on this
* Jim suggested a a unit test for the ReloadableCustomAnalyzer, something that checks that reloading works and that multiple threads will see the update when they create a new token stream. It doesn't need to be focus on synonyms, just checking that the new filters are taken into account when there is an update should be enough.

**Comment by cbuescher** [comment]
@jimczi @romseygeek I added a commit to make both CustomAnalyzer and ReloadableCustomanalyzer implement a common interface that allows getting the AnalyzerComponents in a uniform way. I also added a unit test for ReloadableCustomanalyzer that tests that changes made via reloading become visible to several concurrently running threads. Let me know what you think about those changes and if they address your previous comments.

**Comment by jimczi** [comment]
>  Let me know what you think about those changes and if they address your previous comments.

+1, the change looks good to me. Any idea why the tests are failing ?

**Comment by cbuescher** [comment]
@elasticmachine run  elasticsearch-ci/1

**Comment by cbuescher** [comment]
@jimczi tests are green now, I had to undo a change I did in IndexAnalyzers because it interfered with the (somewhat weird) way we use temporary 
IndexAnalyzers in `MetaDataIndexUpgradeService#checkMappingsCompatibility) and I intend to try improve on that in another issue but this shouldn't
concern this PR. Anything else to add besides getting anohter thumbs up from @romseygeek and/or @jpountz?

File: server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java]: nit: duplicate `the`

File: server/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java]: Can you use the interface (AnalyzerComponentsProvider) that you created to merge the two conditions ?

**Review by jimczi — APPROVED** [review]
I left some minor comments but the change looks good to me. I am also curious to know what @jpountz and @romseygeek think.

File: server/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java]: nit: extra space

File: server/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java]: Can you use AnalysisComponentsProvider here instead? Avoids casting lower down, and makes the instanceof checks simpler.

**Review by romseygeek — APPROVED** [review]
One potential simplification and one nit, LGTM otherwise.

File: server/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java]: +1, I pushed something, needed to change some internal method signatures for it to work but it looks better to me as well. Lets wait for CI to see if I accidentally broke sth., this code is a bit fragile.

**Comment by jpountz** [comment]
I'm a bit under water with various stuff, please don't wait for me to merge it.

**Comment by cbuescher** [comment]
@jpountz sure I'll ask for your review before finally merging the feature branch to master.

**Pull request merged**
